### PR TITLE
[Feature] Complex CUDA Example with a MAT()

### PR DESCRIPTION
**Pull request by PotatoPeeler3000**
Here we have an example of using the JavaCV Mat() that gets passed to a kernel using CUDA and we check that we get the expected result. Again well documented to try and help others learning

### CODE REVIEW

File: ihmc-perception/src/test/java/us/ihmc/perception/cuda/examples/ExampleCUDAKernel2D.java [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+import us.ihmc.perception.cuda.CUDAKernel;
+import us.ihmc.perception.cuda.CUDAProgram;
+import us.ihmc.perception.cuda.CUDAStreamManager;
+
+import java.net.URISyntaxException;
+import java.nio.file.Path;
+import java.util.Objects;
+
+/**
+ * This is an example of using a more complex CUDA kernel that takes in a JavaCV {@link Mat} object. We do this often for kernels, so this example walks you
```
Comment (TomaszTB) [inline ihmc-perception/src/test/java/us/ihmc/perception/cuda/examples/ExampleCUDAKernel2D.java]:
```suggestion
 * This is an example of using a more complex CUDA kernel that takes in an OpenCV {@link Mat} object. We do this often for kernels, so this example walks you
```

I'd say `Mat` is an OpenCV object that we use through JavaCPP, and JavaCV happens to use it as well. Although I kinda see an argument for saying JavaCV. So this is just an idea, not necessarily a request to make the change

File: ihmc-perception/src/test/java/us/ihmc/perception/cuda/examples/ExampleCUDAKernel2D.java [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+{
+   public ExampleCUDAKernel2D() throws URISyntaxException
+   {
+      // We load the kernel from resources; this is a good place to store kernels to separate them from the Java classes
+      Path programPath = Path.of(Objects.requireNonNull(getClass().getResource("matrix_element_wise_addition.cu")).toURI());
+      // Create the stream
+      CUstream_st stream = CUDAStreamManager.getStream();
+
+      // Here we want to maximize the number of threads we can, this helps optimize the kernel for runtime.
+      // By the rules of CUDA, a block cannot have more than 1024 threads.
```
Comment (TomaszTB) [inline ihmc-perception/src/test/java/us/ihmc/perception/cuda/examples/ExampleCUDAKernel2D.java]: We should mention older GPUs are limited to 512 threads per block

File: ihmc-perception/src/test/java/us/ihmc/perception/cuda/examples/ExampleCUDAKernel2D.java [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+            // This creates the inputs to the kernel, as well as the cpu result we expect to get back and use in Java
+            Mat cpuMatA = new Mat(height, width, opencv_core.CV_16UC1, new Scalar(3));
+            Mat cpuMatB = new Mat(height, width, opencv_core.CV_16UC1, new Scalar(5));
+            Mat cpuResult = new Mat();
+
+            // This creates objects that will be allocated on the gpu, these will be used in the kernel since that all happens on the gpu
+            GpuMat gpuMatA = new GpuMat();
+            GpuMat gpuMatB = new GpuMat();
+            GpuMat gpuResult = new GpuMat(cpuMatA.size(), cpuMatA.type());
+            // We need to know the size in bytes and the type of the result we expect to get on the gpu
```
Comment (TomaszTB) [inline ihmc-perception/src/test/java/us/ihmc/perception/cuda/examples/ExampleCUDAKernel2D.java]: Sorry, I never explained this that well. The `size()` method returns a `Size` object that contains the width and height of the `Mat`. Not super directly related to the size in bytes. `type()` returns the data type (as in, 16 bit unsigned shorts). SO internally the constructor DOES calculate byteSize = width * height * bytesPerPixel

File: ihmc-perception/src/test/java/us/ihmc/perception/cuda/examples/ExampleCUDAKernel2D.java [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+         // We need to worry about closing these variables
+         dim3 gridDim = new dim3(); // The same thing as ( new dim3(1, 1, 1); )
+         dim3 blockDim = new dim3(width, height, 1);
+
+         // Runs the kernel with the desired grid and block sizes
+         cudaKernel.run(stream, gridDim, blockDim, 0);
+
+         // Synchronize the stream
+         // This call waits until all asynchronous functions being executed on this stream finish.
+         // We have to call this to ensure that the above memcpy finished, and we have data back in Java land
```
Comment (TomaszTB) [inline ihmc-perception/src/test/java/us/ihmc/perception/cuda/examples/ExampleCUDAKernel2D.java]:
```suggestion
         // We have to call this to ensure that the kernel finished, and the result is ready to be downloaded onto the CPU
```

File: ihmc-perception/src/test/resources/us/ihmc/perception/cuda/examples/matrix_element_wise_addition.cu
Comment (TomaszTB) [inline ihmc-perception/src/test/resources/us/ihmc/perception/cuda/examples/matrix_element_wise_addition.cu:1]: I think we should add comments in this guy as well. I can do it Monday.

File: ihmc-perception/src/test/java/us/ihmc/perception/cuda/examples/ExampleCUDAKernel2D.java (hunk 4 above)
Comment (ds58) [reply ihmc-perception/src/test/java/us/ihmc/perception/cuda/examples/ExampleCUDAKernel2D.java]: I prefer this change as well

File: ihmc-perception/src/test/resources/us/ihmc/perception/cuda/examples/matrix_element_wise_addition.cu
Comment (TomaszTB) [reply ihmc-perception/src/test/resources/us/ihmc/perception/cuda/examples/matrix_element_wise_addition.cu:1]: Done! @PotatoPeeler3000 could you read through it and let me know if anything's too much or too little?

File: ihmc-perception/src/test/java/us/ihmc/perception/cuda/examples/ExampleCUDAKernel2D.java (hunk 3 above)
Comment (PotatoPeeler3000) [reply ihmc-perception/src/test/java/us/ihmc/perception/cuda/examples/ExampleCUDAKernel2D.java]: Ya thanks, sound good

File: ihmc-perception/src/test/resources/us/ihmc/perception/cuda/examples/matrix_element_wise_addition.cu
Comment (PotatoPeeler3000) [reply ihmc-perception/src/test/resources/us/ihmc/perception/cuda/examples/matrix_element_wise_addition.cu:1]: Epic

**Review by TomaszTB — APPROVED** [review]

**Review by ds58 — APPROVED** [review]

**Pull request merged**